lib/data/kinetic/Kinetic.js

```diff
@@ -0,0 +1,360 @@
+import EventEmitter from 'events';
+import net from 'net';
+import stream from 'stream';
+
+import { errors } from 'arsenal';
+import kineticlib from 'kinetic-js';
+
+const HEADER_SZ = 9;
+
+class Kinetic {
+
+    /**
+     * @constructor
+     * @param {Object.<string>[]} drives - the list of drives
+     * @returns {Kinetic} - to allow for a functional style.
+     */
+    constructor(drives) {
+        this._sockets = [];
+        this._drives = drives;
+        this._sockets = this._drives.map(drive => ({
+            drive,
+            clusterVersion: 0,
+            connectionID: 0,
+            sock: this._newSocket(),
+            _request: false,
+            _requestGet: false,
+            _events: new EventEmitter(),
+        }));
+        this._chunkTab = [];
+        this._count = 0;
+        this._lexiKey = '0';
+        this.connect();
+        return this;
+    }
+
+    /**
+     * Connect the sockets to the drives
+     * @return {undefined}
+     */
+    connect() {
+        this._sockets.forEach((socket, index) => {
+            socket.sock.on('connect', () => {
+                this._bindReadable(socket, index);
+            }).connect(socket.drive);
+        });
+    }
+
+    getIt() {
+        for (let i = 0; i < this._sockets.length; i++) {
+            if (this._sockets[i]._request === false
+                && this._sockets[i]._requestGet === false) {
+                return i;
+            }
+        }
+        this._sockets.push(
+            {
+                drive: { host: '127.0.0.1', port: '8123' },
+                clusterVersion: 0,
+                connectionID: 0,
+                sequence: this._sequence,
+                sock: this._newSocket(),
+                _request: false,
+                _requestGet: false,
+                _events: new EventEmitter(),
+            });
+        const index = this._sockets.length - 1;
+        const sock = this._sockets[index];
+        return sock.sock.on('connect', () => {
+            this._bindReadable(sock, index);
+        }).connect(sock.drive);
+    }
+
+    _bindReadable(socket, index) {
+        socket.sock.on('readable', err => {
+            if (socket._requestGet === false) {
+                this._parsePDU(socket, (err1, pdu) => {
+                    this._parseRequest(err || err1, pdu, index);
+                });
+            } else {
+                this._endGet(index);
+            }
+        });
+    }
+
+    _parsePDU(socket, callback) {
+        const header = socket.sock.read(HEADER_SZ);
+        if (header !== null) {
+            const protobufSize = header.readInt32BE(1);
+            const rawData = socket.sock.read(protobufSize);
+            if (rawData !== null) {
+                const pdu = new kineticlib.PDU(
+                    Buffer.concat([header, rawData]));
+                const err = this._propError(pdu);
+                return callback(err, pdu);
+            }
+        }
+        return undefined;
```

Comment:
So in case there is no header, we do not call the callback? That seems a bit stange.